src/crypto/openssl.rs

```diff
@@ -117,35 +124,53 @@ impl From<EcKey<Private>> for OpenSSLLocalKeyPair {
     }
 }
 
-pub struct OpenSSLCrypto;
-impl Crypto for OpenSSLCrypto {
-    type RemotePublicKey = OpenSSLRemotePublicKey;
-    type LocalKeyPair = OpenSSLLocalKeyPair;
+pub struct OpensslCryptographer;
+impl Cryptographer for OpensslCryptographer {
+    fn generate_ephemeral_keypair(&self) -> Result<Box<LocalKeyPair>> {
+        Ok(Box::new(OpenSSLLocalKeyPair::generate_random()?))
+    }
+
+    fn import_key_pair(&self, components: &EcKeyComponents) -> Result<Box<LocalKeyPair>> {
+        Ok(Box::new(OpenSSLLocalKeyPair::from_raw_components(
+            components,
+        )?))
+    }
 
-    fn generate_ephemeral_keypair() -> Result<Self::LocalKeyPair> {
-        Self::LocalKeyPair::generate_random()
+    fn import_public_key(&self, raw: &[u8]) -> Result<Box<RemotePublicKey>> {
+        Ok(Box::new(OpenSSLRemotePublicKey::from_raw(raw)?))
     }
 
     fn compute_ecdh_secret(
-        remote: &Self::RemotePublicKey,
-        local: &Self::LocalKeyPair,
+        &self,
+        remote: &RemotePublicKey,
+        local: &LocalKeyPair,
     ) -> Result<Vec<u8>> {
+        let local_any = local.as_any();
+        let local = local_any.downcast_ref::<OpenSSLLocalKeyPair>().unwrap();
         let private = local.to_pkey()?;
+        let remote_any = remote.as_any();
+        let remote = remote_any.downcast_ref::<OpenSSLRemotePublicKey>().unwrap();
         let public = remote.to_pkey()?;
         let mut deriver = Deriver::new(&private)?;
         deriver.set_peer(&public)?;
         let shared_key = deriver.derive_to_vec()?;
         Ok(shared_key)
     }
 
-    fn hkdf_sha256(salt: &[u8], secret: &[u8], info: &[u8], len: usize) -> Result<Vec<u8>> {
+    fn hkdf_sha256(&self, salt: &[u8], secret: &[u8], info: &[u8], len: usize) -> Result<Vec<u8>> {
         let hk = Hkdf::<Sha256>::extract(Some(&salt[..]), &secret);
         let mut okm = vec![0u8; len];
         hk.expand(&info, &mut okm).unwrap();
         Ok(okm)
     }
 
-    fn aes_gcm_128_encrypt(key: &[u8], iv: &[u8], data: &[u8], tag_len: usize) -> Result<Vec<u8>> {
+    fn aes_gcm_128_encrypt(
+        &self,
+        key: &[u8],
+        iv: &[u8],
+        data: &[u8],
+        tag_len: usize,
```

Comment:
Would this crash if the caller passed in a bogus value for `tag_len`, such as `0`?